discord/rate_limiting.py

```diff
@@ -0,0 +1,465 @@
+        remaining: :class:`int` | `None`
+            The remaining number of requests which can be made
+            to this bucket.
+        reset_after: :class:`int` | `None`
+            When this bucket will next reset.
+        """
+
+        if remaining is None and reset_after is None:
+            self.metadata_unknown = True
+            self.release()
+        else:
+            self.remaining = remaining
+
+            if self._reset_after_set:
+                return
+
+            self._reset_after_set = True
+
+            loop = asyncio.get_running_loop()
+            loop.call_later(cast(float, reset_after), self._reset)
+
+    def _reset(self) -> None:
+        self._reset_after_set = False
+        self.remaining = None
+
+        self.release(self.limit)
+
+
+class BucketStorage:
+    """A customizable, optionally replacable storage medium for buckets.
+
+    Parameters
+    ----------
+    concurrency: :class:`int`
+        The concurrency to reset for every `per` reset.
+    per: :class:`int` | :class:`float`
+        Time to wait until resetting `concurrency.`
+    """
+
+    def __init__(self, per: int = 1, concurrency: int = 50) -> None:
+        self._buckets: dict[str, Bucket] = {}
+        self.global_concurrency = GlobalRateLimit(concurrency, per)
+
+        gc.callbacks.append(self._collect_buckets)
+
+        # temporary buckets, different from normal "permanent" buckets
+        # which cannot be tracked via remaining or reset_after in headers
+        self._temp_buckets: dict[str, DynamicBucket] = {}
+
+    def _collect_buckets(
+        self, phase: Literal["start", "stop"], info: dict[str, int]
+    ) -> None:
+        del info
+
+        if phase == "stop":
+            return
+
+        for id, bucket in self._buckets.copy().items():
+            if bucket.garbage:
+                del self._buckets[id]
+
+    async def append(self, id: str, bucket: Bucket) -> None:
+        """Append a permanent bucket.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+        bucket: :class:`.Bucket`
+            The bucket to append.
+        """
+
+        self._buckets[id] = bucket
+
+    async def get(self, id: str) -> Bucket | None:
+        """Get a permanent bucket.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+
+        Returns
+        -------
+        :class:`.Bucket` or `None`
+        """
+
+        return self._buckets.get(id)
+
+    async def get_or_create(self, id: str) -> Bucket:
+        """Get or create a permanent bucket.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+
+        Returns
+        -------
+        :class:`.Bucket`
+        """
+
+        buc = await self.get(id)
+
+        if buc:
+            return buc
+        else:
+            buc = Bucket()
+            await self.append(id, buc)
+            return buc
```

Comment:
This is redundant. Both paths taken end up returning a bucket.

Reply:
`self.append`?

Reply:
`self.append`?

Reply:
```suggestion
        if not buc:
            buc = Bucket()
            await self.append(id, buc)

        return buc
```
This is what I'm trying to suggest.